Fixes #26760 - Improve performance of the audits page

xprazak2 opened the pull request:
One potential performance bottleneck remains at https://github.com/theforeman/foreman/blob/develop/app/models/concerns/audit_extensions.rb#L25

But since we need to look into all those scopes separately as it fixes [#24232](https://projects.theforeman.org/issues/24232), I am not sure how we could improve it. 



<!--
Simple description of what is fixed/introduced.

Prerequisites for testing:
* VMware cluster
* Two smart proxies

Tests needed (anticipated impacts):
* Hosts list - mainly power column
* Power status on host page
-->

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
* Suggest prerequisites for testing and testing scenarios following example above.
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

-->

lzap commented on db/migrate/20190611104619_add_index_to_audit_type.rb:
```diff
@@ -0,0 +1,9 @@
+  def down
+    remove_index :audits, :auditable_type
+  end
+end
```
We already have this index in:

    t.index ["auditable_type", "auditable_id", "version"], name: "index_audits_on_auditable_type_and_auditable_id_and_version"

The one you want to add was recently reviewed and removed by me and @tbrisker since PostgreSQL can use it. I will need to see a proof that this is the culprit.

lzap replied:
FYI recently there's been reported a performance related bug for audits and this fixed it: https://github.com/theforeman/foreman/pull/6749/files maybe it's related?

tbrisker commented on app/helpers/audits_helper.rb:
```diff
@@ -207,6 +207,11 @@ def nested_host_audit_breadcrumbs
 
   def construct_additional_info(audits)
+    authorized_locations = Location.authorized(:view_locations)
+    current_locations = User.current.my_locations
```
you can unify them using something like:
```
viewable_locations = Location.authorized(:view_locations) | User.current.my_locations
```
Then below would just become 
```
base = audit.locations & viewable_locations
```

tbrisker commented on app/helpers/pagination_helper.rb:
```diff
@@ -16,7 +16,7 @@ def per_page_options(options = [5, 10, 15, 25, 50])
     {
       viewType: 'table',
-      itemCount: collection.total_entries,
+      itemCount: collection.klass.count,
```
Wouldn't this trigger a `COUNT` query on the database every time, as well as ignore the actual collection with the permissions (e.g. in audits this would be Audit.count)?

Commit:
Fixes #26760 - Improve the taxonomix performance

This modifies how with_taxonomy_scope searches
for taxables.

Current behaviour:
1) Find taxable_ids
2) taxable_scope.where(:id => taxable_ids)

New behaviour makes join with taxable_taxonomies
and returns a result of the join.

This prevents fetching large number of records
by id and returns the result of the join directly.

xprazak2 commented:
I managed to track down the cause:
![audit-ids](https://user-images.githubusercontent.com/2664090/61044635-58873f00-a3d9-11e9-92cc-3d643b465067.png)

It turns out this query has performance problems when you have millions of audits. Unfortunately, this requires a change in `Taxonomix`.

Commit:
Refs #26760 - Avoid where in tax scope, fix tests

ares commented:
Seems as test failures are related

Commit:
Refs #26760 - Fix joins

Commit:
Refs #26760 - Fix tests, clean taxable sti

ares commented:
one of the last commit broke many tests :-)

tbrisker commented:
I'm a bit concerned regarding the scope of this change. 
While it may be improving performance of the audits page, it is touching many other areas, including some that are very heavily used in multiple places and could lead to performance regressions or other bugs. It also makes the taxonomix concern, which is already very complex, even more complicated to understand.
Is it possible to somehow fix this performance issue without such a large change?

xprazak2 commented:
@tbrisker, I share your concern, any change in taxonomix has a potential of breaking many things as it cuts across the whole application. I am also worried this area is not 
particularly well tested because recently added graphql tests caught regressions that would have otherwise gone undetected.

The main problem with loading audits is that we need to look into [4 scopes](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/audit_extensions.rb#L25) and each of them fetches the records [by ids](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/taxonomix.rb#L152).

So I removed the lookup by ids and all the subsequent changes were introduced to make the tests pass. I would very much like to have a quick and simple fix without touching taxonomix, but I currently do not see it. I heard we stored the taxonomy information directly in audit, but it had its own problems and we moved away from it, @ares will probably have more details about that.

However the problem with taxonomix performance may hit other resources as well if there is a large number of records: https://projects.theforeman.org/issues/27332

ares commented:
I'd be fine in merging such a big change to taxonomix just after the branching, so we have time to find issues. But I wonder whether we could also simplify this a bit or perhaps drop some functionality of taxonomies or unify it over more resources. If we won't do anything with it, we'll always face these problems. I don't have a good suggestion as I read this PR very quickly, but perhaps enforcing STI models include taxonomix in the parent.

tbrisker commented:
I think we could likely simplify it a bit since it was written at a time where one or both taxonomies could have been disabled. 
Also, do we actually have any resources that are only scoped to one of them using taxonomix? perhaps we should make taxonomix only handle resources that have both, that would also simplify it quite a bit i believe.

xprazak2 commented:
I'll look into how things could be simplified and what could be dropped. 

Jenkins just stopped testing when I push. Maybe he does not want this to be merged :grin:

tbrisker commented:
[test foreman][test katello]

ares commented:
> I think we could likely simplify it a bit since it was written at a time where one or both taxonomies could have been disabled.
Also, do we actually have any resources that are only scoped to one of them using taxonomix? perhaps we should make taxonomix only handle resources that have both, that would also simplify it quite a bit i believe.

Sadly we have all combinations. There are resources that are not taxable at all (Architecture). There are resources taxable only by Org (most of Katello resources). And then all fully taxable we have in core.

Commit:
Refs #26760 - Do not leave scope

tbrisker commented:
> Sadly we have all combinations. There are resources that are not taxable at all (Architecture). There are resources taxable only by Org (most of Katello resources). And then all fully taxable we have in core.

Katello doesn't seem to be using taxonomix at all, and same is true for non-taxable resources. So I would argue that taxonomix should only be used for resources that use both, meaning we can get rid of all the checks for if there is only one of them and assume both are present. That also means we can pull the taxonomy ids from taxable_taxonomies without caring about the taxonomy type for the functions that only need a list of ids.

ares commented:
Oh yeah, for Taxonomix I agree. I thought you were talking about how we handle audits searching, where we have all combinations.

xprazak2 commented:
[test foreman]

Commit:
Refs #26760 - Do not use string conditions on join

tbrisker commented:
I am thinking maybe we are approaching this the wrong way - we are trying to adapt taxonomix to work with audits, which makes it super complicated because some audited resources only have one or no taxonomies, causing a lot of complication for other resources. 
What if instead of that, we change the audit extension to not rely on taxonomix but rather use its own methods that give a better selection of audits that should be displayed?

ares commented:
@tbrisker how would that work? my ultimate use case is, user that is assigned to org A should not see audits for records that are only assigned to org B. And the same should apply to all combinations of taxable, non-taxable, semi-taxable resources.

tbrisker commented:
> @tbrisker how would that work? my ultimate use case is, user that is assigned to org A should not see audits for records that are only assigned to org B. And the same should apply to all combinations of taxable, non-taxable, semi-taxable resources.

Which is exactly why i think the taxonomix methods are not a good fit here - taxonomix should only apply to fully taxable resources. 
I'm trying to come up with an alternative solution, though it is a bit challenging since we have about 72 different ways of handling STI in audits :\

ares commented:
> Which is exactly why i think the taxonomix methods are not a good fit here - taxonomix should only apply to fully taxable resources.

After talking to @xprazak2 I think I now better understand what you suggested. Yeah I agree we don't have to use Taxonomix in Audit model. Or we can but we can override some methods. I'd just suggest you two talk before we double spend the effort :-)

tbrisker commented:
closing in favor of GH-7000
